docs/source/demos/punched_domain.geo.rst

```diff
@@ -0,0 +1,113 @@
+Creating Firedrake compatible meshes in Gmsh
+============================================
+
+The purpose of this demo is to summarize the
+key structure of a ``gmsh.geo`` file that creates
+Firedrake compatible mesh. For more details about Gmsh, please
+refer to the `Gmsh documentation <http://gmsh.info/#Documentation>`_.
+
+The first thing we define are four corners of a rectangle.
+We specify the x,y, and z(=0) coordinates, as well as the target
+element size at these corner (which we set to 0.5). ::
+
+  Point(1) = {-6, 2, 0, 0.5};
+  Point(2) = {-6, -2, 0, 0.5};
+  Point(3) = {6, -2, 0, 0.5};
+  Point(4) = {6, 2, 0, 0.5};
+
+Then, we define 5 points to describe a circle. ::
+
+  Point(5) = {0, 0, 0, 0.1};
+  Point(6) = {0.5, 0, 0, 0.1};
+  Point(7) = {-0.5, 0, 0, 0.1};
+  Point(8) = {0, 0.5, 0, 0.1};
+  Point(9) = {0, -0.5, 0, 0.1};
+
+Then, we create 8 edges: 4 for the rectangle and 4 for the circle.
+Recall that the Gmsh command ``Circle`` requires the arc to be
+strictly smaller than :math:`\pi`. ::
+
+  Line(1) = {1, 4};
+  Line(2) = {4, 3};
+  Line(3) = {3, 2};
+  Line(4) = {2, 1};
+  Circle(5) = {8, 5, 6};
+  Circle(6) = {6, 5, 9};
+  Circle(7) = {9, 5, 7};
+  Circle(8) = {7, 5, 8};
+
+Then, we glue together the rectangle edges and, separately, the circle edges.
+Note that ``Line``, ``Circle``, and ``Curve Loop`` (as well as ``Physical Curve`` below)
+are all curves in Gmsh and must possess a unique tag. ::
+
+  Curve Loop( 9) = {1, 2, 3, 4};
+  Curve Loop(10) = {8, 5, 6, 7};
+
+Then, we define two plane surfaces: the rectangle without the disc first, and the disc itself then. ::
+
+  Plane Surface(1) = {9, 10};
+  Plane Surface(2) = {10};
+
+Finally, we group together some edges and define ``Physical`` entities, which
+can be accessed in Firedrake. ::
+
+  Physical Curve("HorEdges", 11) = {1, 3};
+  Physical Curve("VerEdges", 12) = {2, 4};
+  Physical Curve("Circle", 13) = {8, 7, 6, 5};
+  Physical Surface("PunchedDom", 3) = {1};
+  Physical Surface("Disc", 4) = {2};
+
+For simplicity, we have gathered all this commands in the file
+`punched_domain.geo <punched_domain.geo>`__. To generate a mesh using this file,
+you can type the following command in the terminal::
+
+    gmsh -2 punched_domain.geo -format msh2
```

Comment:
Do we still need the `-format msh2`? I thought this had been fixed in PETSc. @wence?

Reply:
I'll wait for @wence- answers.

Reply:
`-format msh2` is not required. But gmsh 4 is!

Reply:
I removed `-format msh2` and stated clearly that `The Gmsh syntax used in this document is for Gmsh version 4.4.1 .`

Reply:
@dham and @wence- : if I generate the mesh withou `format msh2`, the command `mesh = Mesh('immersed_domain.msh')` gives the error
```
Error: error code 62
[0] DMPlexCreateGmsh() line 1107 in /private/var/folders/tp/07bq8pcj785c7v23xcjqy34r0000gn/T/pip-req-build-wiphcbcn/src/dm/impls/plex/plexgmsh.c
[0] DMPlexCreateGmsh_ReadNodes_v40() line 507 in /private/var/folders/tp/07bq8pcj785c7v23xcjqy34r0000gn/T/pip-req-build-wiphcbcn/src/dm/impls/plex/plexgmsh.c
[0] Invalid argument
[0] Unexpected node number 6 should be 5
```

However, if I include `-forma msh2`, then firedrake does not complain

Reply:
I think if you have `firedrake-update`d, this should now work (at least it did on my computer yesterday).

Reply:
Unfortunately it doesn't work on mine (not even after `firedrake-update`).
What do you suggest to do?

Reply:
I wonder if your gmsh version is too new?

Reply:
Let's just whack `-format msh2` on and be done with it.

Reply:
Probably with a little explanatory text.

Reply:
I've readded it and included a small note.